improver/cli/nowcast_optical_flow_perturbation.py

```diff
@@ -0,0 +1,95 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to calculate optical flow advection velocities with option to
+extrapolate."""
+
+from improver import cli
+
+# Creates the value_converter that clize needs.
+inputadvection = cli.create_constrained_inputcubelist_converter(
+    "precipitation_advection_x_velocity", "precipitation_advection_y_velocity",
+)
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    current_obs: cli.inputcube,
+    previous_forecast: cli.inputcube,
+    previous_advection: inputadvection,
+    orographic_enhancement: cli.inputcube,
+):
+    """Calculate optical flow components from input fields.
+
+    Args:
+        current_obs (iris.cube.Cube):
+            Cube containing latest radar observation
+        previous_forecast (iris.cube.Cube):
+            Cube containing advection nowcast for this time, created at an
+            earlier timestep
+        previous_advection (iris.cube.CubeList):
+            Advection velocities in the x- and y- directions used to generate
+            the previous forecast
+        orographic_enhancement (iris.cube.Cube):
+            Cube containing orographic enhancement for this timestep
+
+    Returns:
+        iris.cube.CubeList:
+            List of the umean and vmean cubes.
+
+    """
+    from iris.cube import CubeList
+
+    from improver.nowcasting.optical_flow import OpticalFlow
+    from improver.nowcasting.utilities import ApplyOrographicEnhancement
+
+    # check that validity times on the forecast and observation match
+    if current_obs.coord("time").cell(0) != previous_forecast.coord("time").cell(0):
+        raise ValueError("Forecast validity time must match input observation")
+
+    # subtract orographic enhancement
+    cube_list = ApplyOrographicEnhancement("subtract")(
+        [current_obs, previous_forecast], orographic_enhancement
+    )
+
+    # calculate optical flow velocity perturbations from previous nowcast to
+    # current observation
+    ofc_plugin = OpticalFlow(iterations=100)
+    advection_perturbations = ofc_plugin(previous_forecast, current_obs, boxsize=30)
+
+    # adjust previous advection components by perturbation values to generate
+    # new components
+    for prev, adj in zip(previous_advection, advection_perturbations):
+        prev.convert_units(adj.units)
+        adj.data += prev.data
```

Comment:
Are the units in advection_perturbations inconsistent across a cube? If not, would it be possible to convert the units all at once, then just sum the cubes in place rather than using for/zip?

Reply:
`advection_perturbations` is a `CubeList` (or list of cubes).  So no, as far as I know there's no way to do this without looping.

Reply:
Ah, so it is. Actually this is better for the memory usage anyway :-)